### PR TITLE
feat(main): Add support for psycopg (Issue #214)

### PR DESCRIPTION
**Pull request by UzielSilva**
See https://github.com/GoogleCloudPlatform/cloud-sql-python-connector/issues/214
Changelog:
- Add proxy for connections that can only be made through an unix socket, to support the TLS connection
- Add support for psycopg, using the proxy server
- Add unit and integration tests
- Update docs

### CODE REVIEW

**Comment by hessjcg** [comment]
Thank you for the contribution! We are excited to add this to the code. I will give this careful review and get back in touch soon with some additional comments.

**Review by hessjcg — COMMENTED** [review]
Nice work so far. This is going to be a very popular enhancement. 

You may want to look at a similar local unix proxy implementation in the [node connector](https://github.com/GoogleCloudPlatform/cloud-sql-nodejs-connector/blob/main/src/connector.ts#L316). 

There are a few things I'd like you to add to this PR. This will get it closer to production-worthy code:

1. The socket path should be a configuration parameter `local_socket_path` on `connector.connect()` instead of being hard-coded. This will allow an application to open many connectors to different database instances.
2. When the listener opens the unix socket, it should set the file permissions to be readable only by the user.
3. Handle exceptions from creating and binding the unix socket, log an appropriate error, and raise a LocalProxyStartupError. Define this new exception in google/cloud/sql/connector/exceptions.py
4. Consider using asyncio instead of a thread to copy data to and from the proxy. 
5. When the connector is closed, the `local_communication` loop must be stopped and the socket file removed.

Again, I'm grateful that you put in the time and wrote this feature. Thank you.

**Comment by enocom** [comment]
cc @rhatgadkar-goog who owns the AlloyDB Python Connector and will be interested to see this.

Big +1 on what @hessjcg. This will be a huge win for people wanting to use psycopg and Cloud SQL (or AlloyDB).

**Comment by UzielSilva** [comment]
All your comments were addressed on my last commit, I even improve the integration test to use SQLAlchemy like other tests.
I only have this issue when I ran the integration test:
```
Task was destroyed but it is pending!
task: <Task pending name='Task-12' coro=<local_communication() done, defined at /home/uziel/Enterprises/TSA/cloud-sql-python-connector/google/cloud/sql/connector/proxy.py:72> wait_for=<Future pending cb=[BaseSelectorEventLoop._sock_read_done(20, handle=<Handle BaseS...0>, 10485760)>)(), Task.task_wakeup()]>>
```
Not sure how to handle or if it is expected, this is caused because the `local_communication` is still expecting your input when you close the connection. Not sure if there is a more friendly way to stop it, but oh well.
Let me know if this is good enough for production.

**Comment by UzielSilva** [comment]
FYI: The last issue I mentioned is only a warning, so, actually, it is not affecting the feature.

**Comment by rhatgadkar-goog** [comment]
> All your comments were addressed on my last commit, I even improve the integration test to use SQLAlchemy like other tests. I only have this issue when I ran the integration test:
> 
> ```
> Task was destroyed but it is pending!
> task: <Task pending name='Task-12' coro=<local_communication() done, defined at /home/uziel/Enterprises/TSA/cloud-sql-python-connector/google/cloud/sql/connector/proxy.py:72> wait_for=<Future pending cb=[BaseSelectorEventLoop._sock_read_done(20, handle=<Handle BaseS...0>, 10485760)>)(), Task.task_wakeup()]>>
> ```
> 
> Not sure how to handle or if it is expected, this is caused because the `local_communication` is still expecting your input when you close the connection. Not sure if there is a more friendly way to stop it, but oh well. Let me know if this is good enough for production.

Thank you for this PR of adding support for psycopg! We really appreciate this change.

I took a look at this StackOverflow post to help understand what might be happening here: https://stackoverflow.com/questions/71542947/how-can-i-fix-task-was-destroyed-but-it-is-pending.

I think you're getting the error, because when the connector closes, you're not cancelling the local proxy server task and waiting for it to shutdown. So maybe you can try making the following changes:
- Modify `start_local_proxy()` function to return the created asyncio task and save this as a member variable in the `Connector` object.
- And then in the `Connector` class's `close_async()` function, you can cancel the proxy server's asyncio task and gracefully wait for it to cancel. [RefreshAheadCache's close() function](https://github.com/GoogleCloudPlatform/cloud-sql-python-connector/blob/28c1c40c7b5e29f858f30d849c8b2c733d623f30/google/cloud/sql/connector/instance.py#L212) might be a good example for how to do this.

**Comment by UzielSilva** [comment]
@rhatgadkar-goog It worked!, now the warning is gone and the server is being stopped gracefully. Thanks!

So, I think the PR is ready for a review, let me know if you want me to do some other modifications. :)

File: google/cloud/sql/connector/connector.py [hunk 1]
```diff
@@ -380,7 +385,7 @@ async def connect_async(
             # async drivers are unblocking and can be awaited directly
             if driver in ASYNC_DRIVERS:
                 return await connector(
-                    ip_address,
```
Comment (kgala2) [inline google/cloud/sql/connector/connector.py:383]: The `host` seems to be undefined here

File: google/cloud/sql/connector/connector.py [hunk 2]
```diff
@@ -390,14 +395,26 @@ async def connect_async(
                 socket.create_connection((ip_address, SERVER_PROXY_PORT)),
                 server_hostname=ip_address,
             )
+
+            host = ip_address
+            # start local proxy if driver needs it
+            if driver in LOCAL_PROXY_DRIVERS:
+                local_socket_path = kwargs.pop("local_socket_path", "/tmp/connector-socket")
+                host = local_socket_path
+                self._proxy = start_local_proxy(
```
Comment (kgala2) [inline google/cloud/sql/connector/connector.py:404]: can we initial the self._proxy in the constructor as it might not be defined for other drivers and will cause error in `close_aync` func

File: google/cloud/sql/connector/psycopg.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+import ssl
+from typing import Any, TYPE_CHECKING
+import threading
```
Comment (kgala2) [inline google/cloud/sql/connector/psycopg.py:19]: This is an unused import, to run and correct lint errors consider - 

pip install nox 
nox -s lint

**Review by kgala2 — CHANGES_REQUESTED** [review]

File: google/cloud/sql/connector/psycopg.py (hunk 3 above)
Comment (UzielSilva) [reply google/cloud/sql/connector/psycopg.py]: All linting issues addressed

File: google/cloud/sql/connector/connector.py (hunk 2 above)
Comment (UzielSilva) [reply google/cloud/sql/connector/connector.py:405]: `self._proxy` variable is now defined as an `Optional[asyncio.Task]` variable in the constructor and initialized as `None`

File: google/cloud/sql/connector/connector.py (hunk 1 above)
Comment (UzielSilva) [reply google/cloud/sql/connector/connector.py]: whoops, moving back to ip_address

**Comment by UzielSilva** [comment]
Also fixed a unit test, test coverage is restored to 100%

**Comment by UzielSilva** [comment]
Trying to add tests for the proxy server, to increase code coverage.

**Comment by UzielSilva** [comment]
Tests were added, code coverage is now at 94%

**Comment by UzielSilva** [comment]
Huh, do I need to also keep support for python 3.9? let me check then the fixes I need to do for it. :)

**Comment by UzielSilva** [comment]
Added support for Python 3.9, only needed to add some exceptions to handle when I close the server

**Comment by UzielSilva** [comment]
Still not able to merge it, it says changes requested

**Comment by enocom** [comment]
Only maintainers can merge.

@kgala2 will have to merge. 

Great work on this @UzielSilva :clap:.

**Comment by kgala2** [comment]
Hi @UzielSilva,

The overall PR looks good, I am checking for corner cases that might not be covered, I'll comment on this PR if I find anything and have psyopg support available as soon as possible. 
cc: @hessjcg 

Thanks

**Comment by sahibpandori** [comment]
Thanks for working on this @UzielSilva, very excited for this change! Hope this can be merged soon @kgala2 @hessjcg

**Comment by UzielSilva** [comment]
Wondering if I should specify the binary flavor for the psycopg optional dependency in the pyproject.toml. I'm using that one for the tests that are working well... not sure how it is going to behave with the purely python based installation.

Ref: https://www.psycopg.org/psycopg3/docs/basic/install.html

CC: @kgala2

File: pyproject.toml [hunk 4]
```diff
@@ -59,6 +59,7 @@ Changelog = "https://github.com/GoogleCloudPlatform/cloud-sql-python-connector/b
 [project.optional-dependencies]
 pymysql = ["PyMySQL>=1.1.0"]
 pg8000 = ["pg8000>=1.31.1"]
+psycopg = ["psycopg>=3.2.9"]
```
Comment (UzielSilva) [inline pyproject.toml:62]: So, maybe we need to update this with `psycopg[binary]`

Comment (jeffclay) [reply pyproject.toml:62]: What are the pro's and con's for the different approaches?

**Review by hessjcg — CHANGES_REQUESTED** [review]
Excellent work so far @UzielSilva. This will fix both #219 and #214 

In reviewing the implementation, this is a good prototype, but it needs to handle sockets in a different way in order to be production-worthy. As implemented today, it can only handle a single connection. [Socket Programming in Python](https://realpython.com/python-sockets/#handling-multiple-connections), particularly the section on "handling multiple connections" explains how the code should handle the unix server socket.

Also, I think we should rename this feature to "Local Unix Socket" instead of "Psychopg support". We are implementing this in a way that is not database driver specific, and will work on any system that supports unix sockets. 

First, lets make a new class in `proxy.py` called `Proxy`. This will encapsulate the details for the proxy server. 

Create a new method on the Connector: `Connector.start_unix_socket_proxy_async`, which will create a new `Proxy` object. The Connector should allow `Connector.start_unix_socket_proxy_async()` to be called multiple times, opening multiple unix sockets as long as the `local_socket_path` parameter does not conflict with an existing socket. 

When the `Proxy` object is created, it will start an "accept loop" async task which will open the unix server socket and listen for new connections. The `Proxy` should keep track of these async tasks so that they can be canceled when `Connector.close()` is called. 

When the "accept loop" task accepts a new connection, a new client socket is opened. Then the accept loop task will create a new async task to handle the client socket. This "client socket" task will do two things:

- call connector.connect() to establish a new socket to the server proxy
- create async tasks to copy data between the unix socket and the database instance socket

The client socket task will also need to handle the following error cases:
- When the `connector.connect()` fails, it should log an error and close the unix socket
- When the database socket is suddenly closed, we should close the unix socket
- When the unix client socket is suddenly closed, we should close the database socket

Cleanup: When connector.close() is called, the connector
   - cancel the "accept loop" task
   - close the unix socket server and delete the socket from the filesystem. 
   - briefly wait for all in-progress async tasks handling connections to close
   - close all open sockets.

I will added  additional comments in-line with the code.

Example of how to use Connector with psychopg and SqlAlchemy

```python
from google.cloud.sql.connector import Connector
async def create_sqlalchemy_engine(
    instance_connection_name: str,
    user: str,
    password: str,
    db: str,
    ip_type: str = "public",
    refresh_strategy: str = "background",
    resolver: Union[type[DefaultResolver], type[DnsResolver]] = DefaultResolver,
) -> tuple[sqlalchemy.engine.Engine, Connector]:
    unix_socket_path="/tmp/conn"

    connector = Connector(refresh_strategy=refresh_strategy, resolver=resolver)
    await connector.start_unix_socket_proxy_async(
        instance_connection_name,
        user=user,
        password=password,
        db=db,
        local_socket_path=unix_socket_path,
        ip_type=ip_type,  # can be "public", "private" or "psc"
        autocommit=True,
    )

    # Construct the connection string
    # Note: The 'host' parameter points to the directory, not the socket file.
    connection_string = f"postgresql+psycopg2://{user}:{password}@{unix_socket_path}/{db}"
    
    # Create the SQLAlchemy engine
    engine = create_engine(connection_string)

    return engine, connector

```

Thank you again for your hard work on this. I look forward to seeing your next revision.
-Jonathan

**Comment by hessjcg** [comment]
Hi @UzielSilva, I was able to spend some time on the tricky IO code to set up a local proxy server. This is still a work in progress, but I wanted to share it anyway. Please check out this branch: 

https://github.com/GoogleCloudPlatform/cloud-sql-python-connector/tree/feat-local-proxy

Currently there are still some failing tests and edge cases that are not properly handled. Please feel free to continue my work.

**Comment by UzielSilva** [comment]
Hi there, sorry for the lack of updates, took some mini vacations.
@hessjcg great stuff!!, I'm moving your proxy server into my branch. This is very helpful. :)

**Comment by UzielSilva** [comment]
I'm having issues trying to fix one integration test, the psycopg one actually.

The issue is about a task that got Future attached to a different loop. The thing is that if I inject the running loop to the connector the issue appears to vanish but the connection now gives us a timeout error. All unit and integration tests are now passing, except for that one, and that is the only issue that is stopping me from finishing this PR. Can I get some help?

**Comment by UzielSilva** [comment]
NVM, I solved it.

@hessjcg I think I'm ready for a new review.

**Comment by seeker25** [comment]
This would be very useful using `pg8000` it's fairly slow especially when it comes to inserting or fetching many rows. I don't think `asyncpg` would be super easy to implement in an existing app that is synchronous. I believe `psycopg` would be much faster.

**Comment by hessjcg** [comment]
Hi @UzielSilva Thank you for the updates! I apologize for the delay. I'll review this PR this week.

**Comment by alberthli** [comment]
Agreed this feature would be really useful! Any updates?

**Comment by JJZolper** [comment]
Yes, definitely hoping for updates soon. I have tried many paths but in Django it is a challenge due to really needing more TCP support. Some of the examples in other threads have given me ideas at least.

**Comment by JJZolper** [comment]
I am definitely hoping to make use of this as soon as it is merged.

**Comment by Jxio** [comment]
Looking forward to this being merged, this looks good would love to use it!

**Comment by milos-blagojevic-tdly** [comment]
Same here, we are hoping to start using this asap!

**Comment by sahibpandori** [comment]
+1, would love to use this once merged!

**Comment by tbCortea** [comment]
any movement on this?

**Comment by seeker25** [comment]
@hessjcg is it possible we could have this looked at please?

**Comment by kgala2** [comment]
Thanks @UzielSilva, I've merged a https://github.com/GoogleCloudPlatform/cloud-sql-python-connector/pull/1452 as part of psycopg support in Python connector, psycopy should be available to use in Python connector's August release!
